### DISCUSSION TITLE
How to show last lines of text in a TextEdit component?

### DISCUSSION THREAD

**Discussion by Argent77** [slint-ui/slint#2760] [Q&A] [answered]
Hi, for a log window I want the TextEdit component to automatically scroll down to the most recent content (i.e. the bottom-most lines) whenever the content changes. How would I realise that with Slint/Rust?

<details>
<summary>Example logger.slint (<a href="https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7B+Button%2C+TextEdit+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+component+Logger+inherits+Window+%7B%0A++++preferred-width%3A+250px%3B%0A++++preferred-height%3A+150px%3B%0A%0A++++VerticalLayout+%7B%0A++++++++Button+%7B%0A++++++++++++private+property+%3Cint%3E+value%3A+1%3B%0A++++++++++++text%3A+%22Add+new+line%22%3B%0A++++++++++++clicked+%3D%3E+%7B%0A++++++++++++++++logger.text+%2B%3D+%22Line+%23%22+%2B+value+%2B+%22+added...%5Cn%22%3B%0A++++++++++++++++value+%2B%3D+1%3B%0A++++++++++++++++logger.edited%28logger.text%29%3B%0A++++++++++++%7D%0A++++++++%7D%0A%0A++++++++logger+%3A%3D+TextEdit+%7B%0A++++++++++++read-only%3A+true%3B%0A++++++++++++edited+%3D%3E+%7B%0A++++++++++++++++%2F%2F+update+viewport+position%3F%3F%3F%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A">SlintPad version</a>)

</summary>

```js
import { Button, TextEdit } from "std-widgets.slint";

export component Logger inherits Window {
    preferred-width: 250px;
    preferred-height: 150px;

    VerticalLayout {
        Button {
            private property <int> value: 1;
            text: "Add new line";
            clicked => {
                logger.text += "Line #" + value + " added...\n";
                value += 1;
                logger.edited(logger.text);
            }
        }

        logger := TextEdit {
            read-only: true;
            edited => {
                // update viewport position???
            }
        }
    }
}
```

</details>

**Comment by ogoffart** [comment, marked as answer]
You can use the viewport property of the TextEdit:
```
                logger.viewport-y = logger.visible-height - logger.viewport-height;
```

I also recommand making a new function rather than re-using the edited callback,  but you can still use the edited callback if you like

[SlintPad](https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7B+Button%2C+TextEdit+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+component+Logger+inherits+Window+%7B%0A++++preferred-width%3A+250px%3B%0A++++preferred-height%3A+150px%3B%0A%0A++++VerticalLayout+%7B%0A++++++++Button+%7B%0A++++++++++++private+property+%3Cint%3E+value%3A+1%3B%0A++++++++++++text%3A+%22Add+new+line%22%3B%0A++++++++++++clicked+%3D%3E+%7B%0A++++++++++++++++logger.text+%2B%3D+%22Line+%23%22+%2B+value+%2B+%22+added...%5Cn%22%3B%0A++++++++++++++++value+%2B%3D+1%3B%0A++++++++++++++++logger.to-end%28%29%0A++++++++++++%7D%0A++++++++%7D%0A%0A++++++++logger+%3A%3D+TextEdit+%7B%0A++++++++++++read-only%3A+true%3B%0A++++++++++++function+to-end%28%29+%7B%0A++++++++++++++++logger.viewport-y+%3D+logger.visible-height+-+logger.viewport-height%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A)

**Comment by Argent77** [comment]
That works great, thanks.
